polyfill/lib/ecmascript.mjs

```diff
@@ -3316,182 +3276,128 @@ export function NanosecondsToDays(nanoseconds, zonedRelativeTo, timeZoneRec, pre
+      milliseconds %= 1000;
+      minutes = MathTrunc(seconds / 60);
+      seconds %= 60;
+      hours = MathTrunc(minutes / 60);
+      minutes %= 60;
+      days = MathTrunc(hours / 24);
+      hours %= 24;
       break;
     case 'hour':
-      ({ quotient: microseconds, remainder: nanoseconds } = nanoseconds.divmod(1000));
-      ({ quotient: milliseconds, remainder: microseconds } = microseconds.divmod(1000));
-      ({ quotient: seconds, remainder: milliseconds } = milliseconds.divmod(1000));
-      ({ quotient: minutes, remainder: seconds } = seconds.divmod(60));
-      ({ quotient: hours, remainder: minutes } = minutes.divmod(60));
+      microseconds = MathTrunc(nanoseconds / 1000);
+      nanoseconds %= 1000;
+      milliseconds = MathTrunc(microseconds / 1000);
+      microseconds %= 1000;
+      seconds += MathTrunc(milliseconds / 1000);
+      milliseconds %= 1000;
+      minutes = MathTrunc(seconds / 60);
+      seconds %= 60;
+      hours = MathTrunc(minutes / 60);
+      minutes %= 60;
       break;
     case 'minute':
-      ({ quotient: microseconds, remainder: nanoseconds } = nanoseconds.divmod(1000));
-      ({ quotient: milliseconds, remainder: microseconds } = microseconds.divmod(1000));
-      ({ quotient: seconds, remainder: milliseconds } = milliseconds.divmod(1000));
-      ({ quotient: minutes, remainder: seconds } = seconds.divmod(60));
+      microseconds = MathTrunc(nanoseconds / 1000);
+      nanoseconds %= 1000;
+      milliseconds = MathTrunc(microseconds / 1000);
+      microseconds %= 1000;
+      seconds += MathTrunc(milliseconds / 1000);
+      milliseconds %= 1000;
+      minutes = MathTrunc(seconds / 60);
+      seconds %= 60;
       break;
     case 'second':
-      ({ quotient: microseconds, remainder: nanoseconds } = nanoseconds.divmod(1000));
-      ({ quotient: milliseconds, remainder: microseconds } = microseconds.divmod(1000));
-      ({ quotient: seconds, remainder: milliseconds } = milliseconds.divmod(1000));
+      microseconds = MathTrunc(nanoseconds / 1000);
+      nanoseconds %= 1000;
+      milliseconds = MathTrunc(microseconds / 1000);
+      microseconds %= 1000;
+      seconds += MathTrunc(milliseconds / 1000);
+      milliseconds %= 1000;
       break;
     case 'millisecond':
-      ({ quotient: microseconds, remainder: nanoseconds } = nanoseconds.divmod(1000));
-      ({ quotient: milliseconds, remainder: microseconds } = microseconds.divmod(1000));
+      microseconds = MathTrunc(nanoseconds / 1000);
+      nanoseconds %= 1000;
+      milliseconds = MathTrunc(microseconds / 1000) + seconds * 1000;
+      microseconds %= 1000;
+      seconds = 0;
       break;
     case 'microsecond':
-      ({ quotient: microseconds, remainder: nanoseconds } = nanoseconds.divmod(1000));
+      microseconds = MathTrunc(nanoseconds / 1000) + seconds * 1e6;
+      nanoseconds %= 1000;
+      seconds = 0;
       break;
     case 'nanosecond':
+      nanoseconds += seconds * 1e9;
```

Comment:
This is also imprecise, but can easily be implemented in C++ using `std::fma`. (Same applies for the `'microseconds'` case.)

For example `8692288669465520 * 1_000_000_000 + 321_414_345` is `8692288669465519765585920` when using Number math, but the exact result is `8692288669465520839327744`.

Reply:
I've added a similar note in the spec about this.